feat: Allow exporting data for SFT, Reward Modelling (related to RLHF), DPO, rename TrainingTaskMapping

tomaarsen opened the pull request:
Resolves #3379, resolves #3377
Closes #3522 

Hello!

## Pull Request overview
* Prepare data for SFT, RM, DPO in TRL.
* Rename `TrainingTaskMapping` to `TrainingTask` and `task_mapping` to `task`.

# Description
## Prepare data
```python
from argilla.feedback import TrainingTask

def formatting_func(sample: Dict[str, Any]):
    ...
    yield template.format(
        prompt=sample["prompt"],
        response=sample["response"],
    )

task = TrainingTask.for_supervised_fine_tuning(formatting_func=formatting_func)
ds = fds_dataset.prepare_for_training(framework="trl", task=task)
# -> ds has "text" and "id" columns
```
Compatible with [SFTTrainer](https://huggingface.co/docs/trl/main/en/sft_trainer).

```python
task = TrainingTask.for_reward_modelling(chosen_rejected_func=chosen_rejected_func)
ds = fds_dataset.prepare_for_training(framework="trl", task=task)
# -> ds has "chosen" and "rejected" columns
```
Nearly compatible with [RewardTrainer](https://huggingface.co/docs/trl/main/en/reward_trainer).

```python
task = TrainingTask.for_direct_preference_optimization(prompt_chosen_rejected_func=prompt_chosen_rejected_func)
ds = fds_dataset.prepare_for_training(framework="trl", task=task)
# -> ds has "prompt", "chosen" and "rejected" columns
```
Compatible with [DPOTrainer](https://huggingface.co/docs/trl/main/en/dpo_trainer).

### Details
I implement this by calling `dataset.format_as("datasets")` and then passing each sample (a simple dictionary) from this dataset to the function that the user provides. This user provided function can return `None`, one sample, a list of samples, or yield samples. This allows users to export multiple training samples from a single Argilla record, e.g. when there's multiple annotators that provided useful corrections, or if the annotated record justifies 3 "chosen", "rejected" pairs because there's a ranking between 3 texts.

## Rename
`TrainingTaskMapping` is now `TrainingTask` - the "mapping" part is just unintuitive to the user. Same for `task_mapping` to `task`. **Note:** If people used `task_mapping=...` before, that will now fail. I can make this deprecation softer, but then I have to make `task` optional, which I would rather not do.

## TODO:
- [ ] Add TRL to `ArgillaTrainer`, allowing:
   ```python
   task = TrainingTask.for_supervised_fine_tuning(
       formatting_func=formatting_func
   ) # or any other task from this PR
   trainer = ArgillaTrainer(
       dataset=fds_dataset,
       task=task,
       framework="trl",
   )
   trainer.train()
   ```
- [ ] Consider renaming `FeedbackDataset.prepare_for_training` to `FeedbackDataset.export`.
- [ ] New tests
- [ ] Add documentation

**Type of change**

- [x] New feature

**How Has This Been Tested**

Not finished yet.

**Checklist**

- [ ] I added relevant documentation
- [x] follows the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

---

- Tom Aarsen

tomaarsen committed:
Prevent error if no annotated data exists

tomaarsen committed:
Set 'id' as integers instead of string 'None'

tomaarsen committed:
Fix broken docstrings

tomaarsen committed:
Add RankingStrategy and Unification to __init__

tomaarsen committed:
Allow preparing data for Supervised Finetuning w. TRL

tomaarsen committed:
Use formatting_func instead for preparing SFT

tomaarsen committed:
Rename 'TrainingTaskMapping...' to 'TrainingTask...'

tomaarsen committed:
Rename 'TrainingTaskMapping' to 'TrainingTask' in docs

tomaarsen committed:
Rename `task_mapping` parameter to `task`

No deprecation strategy in place here, people's code might fail if they use
task_mapping=...

tomaarsen committed:
Update incorrect task type hint

tomaarsen committed:
Merge branch 'develop' of https://github.com/argilla-io/argilla into integration/trl_sft

tomaarsen committed:
Prepare data for reward modelling

tomaarsen committed:
Support preparing data for DPO

tomaarsen committed:
Set up initial ArgillaTRLTrainer skeleton

tomaarsen committed:
Implement SFTTrainer, RewardTrainer, DPOTrainer

tomaarsen committed:
Allow updating configs, add __repr__

tomaarsen committed:
Prevent predict from being used

tomaarsen committed:
Merge branch 'develop' of https://github.com/argilla-io/argilla into integration/trl_sft

tomaarsen committed:
When there's no annotations, use a dictionary rather than empty

This is easier to parse for users writing formatting functions etc.

tomaarsen committed:
Add initial tests for preparing data for the 3 TRL tasks

tomaarsen committed:
Prevent crash when using RM/DPO with the old Mapping

tomaarsen committed:
Add train_size to all TRL tests

tomaarsen committed:
Add deprecation tests

tomaarsen committed:
Add extra cases: returning None, one sample, multiple samples

tomaarsen committed:
Add ArgillaTrainer tests

tomaarsen commented on src/argilla/client/feedback/integrations/huggingface/dataset.py:
```diff
@@ -112,12 +112,17 @@ def _huggingface_format(dataset: "FeedbackDataset") -> "Dataset":
             for question in dataset.questions:
                 if not record.responses:
-                    hf_dataset[question.name].append(None)
+                    hf_dataset[question.name].append(empty)
```
@alvarobartt Thoughts on this?

Otherwise users need None checks everywhere, and now they can just iterate over all elements in `sample["value"]`.

By the way, this is relevant for the TRL tasks. See for example the tests in this PR.

tomaarsen replied:
Additionally, the current approach is "columnar", e.g.:
```python
{
      "user_id": [..., ..., ...],
      "value": [4, 3, 1],
      "status": ["submitted", "submitted", "discarded"],
}
```
But have we considered a row-based approach?
```python
[{"user_id": ..., "value": 4, "status": "submitted"},
{"user_id": ..., "value": 3, "status": "submitted"},
{"user_id": ..., "value": 1, "status": "discarded"}]
```
With the latter setup, it is much easier to select all values with the right status:
```python
[annotation["value"] for annotation in X if annotation["status"] == "submitted"]
```
where X is the list.

Right now, you have to do something like:
```python
[value for value, status in zip(X["value"], X["status"]) if status == "submitted"]
```

Thoughts?

alvarobartt replied:
Yes I agree with a row-approach, I also think is better, we would lose consistency in case we want to add a non-matching dict, but I think that even though the `datasets.Features` may be more generic if we rely on Python dicts rather than on `datasets.Sequence`, it's probably better and easier to handle this way

tomaarsen committed:
Add trl to dev deps

tomaarsen committed:
merge 'develop' into integration/trl_sft

tomaarsen committed:
Merge branch 'develop' into integration/trl_sft

tomaarsen committed:
Update finetuning docs, including :meth: link

tomaarsen committed:
Remove :meth:, not worth the time for now

tomaarsen committed:
Fix dunder repr tuples

alvarobartt commented on docs/_source/_common/snippets/training/feedback-task/text-classification.md:
```diff
@@ -1,6 +1,6 @@
 ---
 title: Text classification
-description: When a RatingQuestion, LabelQuestion or MultiLabelQuestion is present in the datasets, we can define a TrainingTaskMappingForTextClassification to use our ArgillaTrainer integration for fine-tuning with openai”, “setfit”, “peft”, “spacy” and “transformers”.
+description: When a RatingQuestion, LabelQuestion or MultiLabelQuestion is present in the datasets, we can define a TrainingTaskForTextClassification to use our ArgillaTrainer integration for fine-tuning with openai”, “setfit”, “peft”, “spacy” and “transformers”.
```
Replaced Windows double-quotes with standard double-quotes, and add missing one in `openai`

```suggestion
description: When a RatingQuestion, LabelQuestion or MultiLabelQuestion is present in the datasets, we can define a TrainingTaskForTextClassification to use our ArgillaTrainer integration for fine-tuning with "openai", "setfit", "peft", "spacy" and "transformers".
```

tomaarsen replied:
Are these Windows double quotes? I don't use them myself. Thanks for the fix.

alvarobartt commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -306,7 +311,12 @@ def unify_responses(
+        task: Union[
+            TrainingTaskForTextClassification,
+            TrainingTaskForSupervisedFinetuning,
+            TrainingTaskForRewardModelling,
```
As the user may directly interact with e.g. `TrainingTask.for_text_classification`, shouldn't we shorten these names?

e.g. `TrainingTaskForDirectPreferenceOptimization` -> `TrainingTaskForDPO`

tomaarsen replied:
I'm down for this. @davidberenstein1957 how do you feel about it? We'd have DPO, SFT, RM, (and later PPO).

alvarobartt commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -306,7 +311,12 @@ def unify_responses(
+            TrainingTaskForRewardModelling,
+            TrainingTaskForDirectPreferenceOptimization,
+        ],
         train_size: Optional[float] = 1,
```
I thought the standard was to define the `Optional` on fields that are `None` by default, as Python will mark those as optional internally if those have a predefined value as `float = 1`, so I would change this to

```suggestion
        train_size: float = 1.0,
```

tomaarsen replied:
`Optional[X]` just means `Union[None, X]`. So, this says `Union[None, float]`. When looking at the code, I think that's true. E.g. the `_prepare_for_training_with_transformers` accepts None as a valid parameter.

alvarobartt commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -376,11 +393,15 @@ def prepare_for_training(
                     lang = spacy.load(lang)
-            return task_mapping._prepare_for_training_with_spacy(data=data, train_size=train_size, seed=seed, lang=lang)
+            return task._prepare_for_training_with_spacy(data=data, train_size=train_size, seed=seed, lang=lang)
         elif framework is Framework.SPARK_NLP:
-            return task_mapping._prepare_for_training_with_spark_nlp(data=data, train_size=train_size, seed=seed)
+            return task._prepare_for_training_with_spark_nlp(data=data, train_size=train_size, seed=seed)
```
IMO we could safely drop `spaCy` and `SparkNLP` support, right? If we want to keep those for a strong reason we can discuss about it

tomaarsen replied:
I think it should be fine to remove once we remove/replace the old trainers. The spaCy transformer tests actually cause my pytest to abruptly terminate. Not ideal.

alvarobartt commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -349,19 +359,26 @@ def prepare_for_training(
+        ):
+            raise ValueError(f"Training data {type(task)} is not supported yet")
+
+        data = task._format_data(self)
```
I would like to simplify the code below this line, when checking for the frameworks and the methods required to prepare the data

tomaarsen replied:
👍 Do you want to do that as a follow-up PR?

alvarobartt replied:
Sure, we can have a look at it upon merge! 👍🏻

alvarobartt commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -48,7 +59,7 @@ def __init__(
         Args:
             dataset (FeedbackDataset): the dataset to be used for training.
-            task_mapping (TrainingTaskMappingForTextClassification): the training data to be used for training.
+            task (TrainingTask): the training data to be used for training.
```
I feel we could just remove the type from the docstrings i.e. `task (TrainingTask):` as it's already defined in the `__init__` so it feels a little bit redundant IMO, thoughts? @tomaarsen @gabrielmbmb

alvarobartt replied:
Also we're not really consistent with docstrings here and there so we should probably spend some time defining our own guidelines for docstrings 👍🏻

tomaarsen replied:
I'm okay with removing it. `TrainingTask` is kind of wrong anyways, as for some reason `TrainingTask` isn't a superclass of `TrainingTaskFor...`.

tomaarsen committed:
Update RLHF reward modelling tutorial

alvarobartt commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -219,24 +240,31 @@ class ArgillaTrainerSkeleton(ABC):
     def __init__(
         self,
         feedback_dataset: "FeedbackDataset",
```
Maybe?

```suggestion
        dataset: "FeedbackDataset",
```

tomaarsen replied:
Done

alvarobartt commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -219,24 +240,31 @@ class ArgillaTrainerSkeleton(ABC):
+            TrainingTaskForTextClassification,
+            TrainingTaskForSupervisedFinetuning,
+            TrainingTaskForRewardModelling,
+            TrainingTaskForDirectPreferenceOptimization,
```
I've seen these block a couple times already, can't we just define it as `TrainingTaskTypes = Union[TrainingTaskForTextClassification, TrainingTaskForSupervisedFinetuning, TrainingTaskForRewardModelling, TrainingTaskForDirectPreferenceOptimization]`?

tomaarsen replied:
Perhaps the best play is just to have one common superclass of all of these that we can use, but I'll go with this approach for now

alvarobartt commented on src/argilla/client/feedback/training/frameworks/transformers.py:
```diff
@@ -57,9 +57,9 @@ def __init__(self, *args, **kwargs):
             self._model_class = AutoModelForSequenceClassification
         else:
-            raise NotImplementedError(f"ArgillaTransformersTrainer does not support {type(self._task_mapping)} yet.")
+            raise NotImplementedError(f"ArgillaTransformersTrainer does not support {type(self._task)} yet.")
```
Maybe?

```suggestion
            raise NotImplementedError(f"ArgillaTransformersTrainer does not support {self._task.__name__} yet.")
```

tomaarsen replied:
I think there's a chance that this bit is dead code, but I'll use `self._task.__class__.__name__`

alvarobartt commented on src/argilla/feedback/__init__.py:
We should probably refactor everything into `src/argilla/feedback/...` instead, rather than under `src/argilla/client/feedback/...`, as I guess that was initially suggested by @frascuchon and @davidberenstein1957 created this file temporarily to "ease" the imports. But maybe we can discuss the refactoring of the module where everything's placed in the upcoming release?

tomaarsen replied:
Sounds good. I'll leave that for after this PR.

alvarobartt commented on src/argilla/training/base.py:
```diff
@@ -281,7 +281,7 @@ def predict(self, text: Union[List[str], str], as_argilla_records: bool = True,
         return self._trainer.predict(text=text, as_argilla_records=as_argilla_records, **kwargs)
 
-    def train(self, output_dir: str = None):
+    def train(self, output_dir: str):
```
```suggestion
    def train(self, output_dir: str) -> None:
```

alvarobartt replied:
Also, is the `output_dir` mandatory?

tomaarsen replied:
Yes. It's mandatory in `transformers` and our ArgillaTrainer `transformers` integration too.

alvarobartt reviewed (commented):
LGTM overall, we can discuss about this PR in detail tomorrow! I still need to review some files, but so far so good, great work Tom!

tomaarsen committed:
Use dunder class dunder name instead of type()

tomaarsen committed:
Use TrainingTaskTypes shorthand

tomaarsen committed:
Rename feedback_dataset to dataset

tomaarsen committed:
Fix weird quotes & '...TaskMapping...'

Co-authored-by: Alvaro Bartolome <alvarobartt@gmail.com>

tomaarsen committed:
Rename internal classes to abbrevations (SFT/RM/DPO)

tomaarsen committed:
Fix broken docstring text

tomaarsen committed:
Remove types from docstring as suggested

tomaarsen committed:
Remove unused 'new' argument, add return type hints

tomaarsen committed:
Reintroduce 'new' keyword, mimicking the transformers trainer

tomaarsen committed:
Increment dev transformers version for DPO

tomaarsen committed:
Remove AutoTrain as suggested by David

tomaarsen committed:
Add TRL to train_prepare_for_training

It doesn't really work, as you need to specify a task, right?

tomaarsen committed:
Add TRL to train_update_config

alvarobartt committed:
fix: add missing `id` in question `Features`

alvarobartt committed:
fix: add `metadata` feature in any case

alvarobartt committed:
feat: row-based dump of `responses`

alvarobartt committed:
fix: `exclude` fields when `RemoteFeedbackRecord`

alvarobartt committed:
chore: use `token` instead of `use_auth_token`

alvarobartt committed:
fix: ensure back comp in `from_huggingface`

alvarobartt committed:
fix: `suggestions` no longer skipped

tomaarsen committed:
Fix critical bug in from_huggingface loading

tomaarsen committed:
Merge pull request #2 from argilla-io/feat/hf-datasets-dict-features

feat: `datasets.Dataset` to export `responses` as `List[dict]` instead of `Sequence`

tomaarsen committed:
Update documentation, docstrings and tests to row-wise format

tomaarsen committed:
Merge branch 'develop' of https://github.com/argilla-io/argilla into integration/trl_sft

tomaarsen committed:
Rename TrainingTaskMapping to TrainingTask in test

davidberenstein1957 committed:
chore: added initial version of PPOTrainer support (#3549)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

I added support for the PPOTrainer.

```python
from transformers import pipeline
from trl import PPOConfig

task_mapping = TrainingTask.for_proximal_policy_optimization(text=dataset.field_by_name("text"))
trainer = ArgillaTrainer(
    dataset=dataset,
    task=task_mapping,
    framework="trl",
    fetch_records=False
)
# assuming we have an arbitrarily trained textcat model
reward_model = pipeline("sentiment-analysis", model="lvwerra/distilbert-imdb")
trainer.update_config(reward_model=sentiment_pipe) # this is always required but if not done it provides a warning
trainer.train(output_dir="my_awesone_model")
```

Closes #3522 

**Type of change**

- [X] New feature (non-breaking change which adds functionality)
- [X] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

- [ ] `tests/integration/client/feedback/training/test_trl.py`

**Checklist**

- [ ] I added relevant documentation
- [X] I followed the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>
Co-authored-by: Tom Aarsen <Cubiegamedev@gmail.com>

tomaarsen committed:
Update Reward Modeling docs

tomaarsen committed:
Improve DPO training with batching & truncation

tomaarsen committed:
Wrap up RLHF documentation

tomaarsen committed:
Showcase that SFT works in docs

tomaarsen committed:
Improve formatting_func for TRL + PPO

dvsrepo commented on docs/_source/guides/llms/practical_guides/fine_tune.md:
```diff
@@ -71,54 +71,159 @@ import argilla as rg
-####  TRL
+#### TRL
 
 The [Transformer Reinforcement Learning (TRL)](https://huggingface.co/docs/trl) package provides a flexible and customizable framework for fine-tuning models. It allows users to have fine-grained control over the training process, enabling them to define their functions and to further specify the desired behavior of the model. This approach requires a deeper understanding of reinforcement learnin ...
```
```suggestion
The [Transformer Reinforcement Learning (TRL)](https://huggingface.co/docs/trl) package provides a flexible and customizable framework for fine-tuning models. It allows users to have fine-grained control over the training process, enabling them to define their functions and to further specify the desired behavior of the model. This approach requires a deeper understanding of reinforcement learning concepts and techniques, as well as more careful experimentation. It is best suited for users who have experience in reinforcement learning and want fine-grained control over the training process. Additionally, it directly integrates with [Parameter-Efficient Fine-Tuning](https://huggingface.co/docs/peft/index) (PEFT) decreasing the computational complexity of this step of training an LLM.
```

dvsrepo commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -33,7 +34,7 @@ class ArgillaTrainer(ArgillaTrainerV1):
+        task: TrainingTaskTypes,
         framework: Framework,
         lang: Optional["spacy.Language"] = None,
         model: Optional[str] = None,
```
This a cross-cutting issue, not coming from this PR, but it's specially relevant for LLMs:

Allowing users to pass only a `str` model ref is inflexible. Very often you want to configure your model with certain parameters during model instantiation (e.g., `AutoModelForCausalLM.from_pretrained`). Enabling users to pass a model (and tokenizer) object would result in a much more flexible API. 

Another potential solution would be to define a new arg like the (trainer) `framework_kwargs` but for model (and tokenizer). But I think this is less desirable in terms of developer experience and maintenance from our side.

Another alternative is to tell our users: if you want this level of flexibility just use `prepare_for_training` and do the training directly with trl, transformers, etc. But I think, if feasible, we should try to give this extra-level of flexibility even if ArgillaTrainer is a short-cut for less critical/customized training workflows.

A clear example of why users will need more flexibility is quantization (needed to train larger models with limited compute resources):

```python
compute_dtype = getattr(torch, bnb_4bit_compute_dtype)
bnb_config = BitsAndBytesConfig(
    load_in_4bit=use_4bit,
    bnb_4bit_quant_type=bnb_4bit_quant_type,
    bnb_4bit_compute_dtype=compute_dtype,
    bnb_4bit_use_double_quant=use_nested_quant,
)

model = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-2-7b-hf",
    quantization_config=bnb_config,
    device_map=device_map
)

```

To the best of my knowledge, using this "configured" `model` is not possible with the current `ArgillaTrainer`. I'm aware this comes from the original ArgillaTrainer API, but maybe it's a good opportunity to tackle this now?

As an additional comment, I believe passing a instantiated model would be relevant for `spacy` too (haven't checked the ArgillaSpacyTrainer so I'm guessing it's the same but might be not):

```python

def custom_tokenizer(nlp):
   ....
   return Tokenizer(....)

nlp = spacy.load('en')
nlp.tokenizer = custom_tokenizer(nlp)

```
Love to hear your thoughts @davidberenstein1957 @tomaarsen @alvarobartt

dvsrepo replied:
Any thoughts?

tomaarsen replied:
I intended to review this with Alvaro and David today, but I didn't get around to it.

I think it should be feasible to allow either a model string or an initialized model. We would then also expose an optional tokenizer parameter. I think this should only require minimal changes, although I think it's best to create these changes in a separate PR.

dvsrepo replied:
Great, I agree with creating a separate PR for this.

tomaarsen committed:
Fix PEFT abbreviation

Co-authored-by: Daniel Vila Suero <daniel@argilla.io>

tomaarsen committed:
Truncate tokenization

tomaarsen committed:
Remove samples that are too long

tomaarsen committed:
Merge branch 'develop' of https://github.com/argilla-io/argilla into integration/trl_sft

tomaarsen committed:
Merge branch 'integration/trl_sft' of https://github.com/tomaarsen/argilla into integration/trl_sft

tomaarsen committed:
Specify formatting function output type hint

tomaarsen committed:
Add missing imports in codeblock

davidberenstein1957 committed:
feat: add `formatting_func` support for TextClassification (#3599)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

Added `formatting_func` support for TextClassification. This should
allow us to pass either `formatting_func` or `text-label` pair to the
task definition.

docs: update broken python formatting fune_tune.md
docs: added content for `formatting_func` for text_classification
chore: added `formatting_func` support for TextClassification
tests: chore: added `formatting_func` tests for TextClassification

Closes #3540 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR
according to the type of change)

- [X] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

This was tested using a text_classification_with_formatting_func  test

**Checklist**

- [X] I added relevant documentation
- [X] I followed the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [X] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: Alvaro Bartolome <alvaro@argilla.io>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

alvarobartt commented on tests/integration/client/feedback/training/test_trainer.py:
```diff
@@ -107,3 +103,33 @@ def test_prepare_for_training_text_classification(
+        lambda: TrainingTaskMapping.for_supervised_fine_tuning(None),
+    ),
+)
+def test_deprecations(callable) -> None:
```
```suggestion
def test_deprecations(callable: Callable[..., Any]) -> None:
```

alvarobartt replied:
Besides that you probably also need to include `from typing import Callable`

tomaarsen replied:
e677447831dc9d7dd53abff2bdfed7934b237a40

alvarobartt commented on environment_dev.yml:
```diff
@@ -46,14 +46,14 @@ dependencies:
       - https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.5.0/en_core_web_sm-3.5.0.tar.gz
       - spacy-transformers>=1.2.5
-      - transformers[torch]>=4.19.0
+      - transformers[torch]>=4.30.0 # <- required for DPO with TRL
```
Make sure to also include those dependencies in `pyproject.toml`

tomaarsen replied:
f357c53240243555bc8818d38eb799c0c53298f6

alvarobartt commented on environment_dev.yml:
```diff
@@ -46,14 +46,14 @@ dependencies:
       - openai
       - peft
-      - autotrain-advanced==0.5.2
+      - trl>=0.5.0
```
Same here, please include it in `pyproject.toml`

tomaarsen replied:
f357c53240243555bc8818d38eb799c0c53298f6

alvarobartt commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -306,14 +313,27 @@ def unify_responses(
+            test_size: the size of the test set. If `None`, the whole dataset will be used for testing.
+            seed: the seed to use for splitting the dataset into train and test sets.
+            lang: the spaCy language to use for training. If `None`, the language of the dataset will be used.
+            fetch_records: whether to fetch the records from Argilla or use the local records instead. If `None`, use local.
+        """
```
`fetch_records` will be deprecated from 1.15.0 onwards, so maybe we can just remove this and iterate over `dataset.records`, since it will fetch those from Argilla if not there, or just loop over those

alvarobartt replied:
```suggestion
        """
```

davidberenstein1957 replied:
@alvarobartt I agree, this aligns with the mention I had regarding the differentiation between local and remote datasets.

tomaarsen replied:
I'll delete it from the ArgillaTrainer. Should I also remove it from `prepare_for_training`? I think that might break backwards compatibility.

tomaarsen replied:
2b576dfa6be51eaa106c96623a56f37fbea9de06

tomaarsen replied:
As discussed privately in Slack, we'll keep this for a separate deprecation PR.

alvarobartt commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -306,14 +313,27 @@ def unify_responses(
+        """
         if fetch_records is not None:
             warnings.warn(
                 "`fetch_records` is deprecated and will be removed in a future version."
```
I'd say here you can just go ahead and remove it, since it is scoped from 1.15.0, so there's no need to still keep this code-block

tomaarsen replied:
As discussed privately in Slack, we'll keep this for a separate deprecation PR.

alvarobartt commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -263,13 +271,13 @@ def predict(self, text: Union[List[str], str], as_argilla_records: bool = True,
 
     @abstractmethod
-    def train(self, output_dir: str = None):
+    def train(self, output_dir: str = None) -> None:
```
```suggestion
    def train(self, output_dir: Optional[str] = None) -> None:
```

alvarobartt commented on tests/integration/client/feedback/training/test_trainer.py:
```diff
@@ -43,8 +45,8 @@
-        Framework("spacy"),
-        Framework("spacy-transformers"),
+        # Framework("spacy"),
+        # Framework("spacy-transformers"),
```
Shouldn't we just remove those if not ran?

```suggestion
```

tomaarsen replied:
Oh, oops. I think these should run as they work on the CI - they just cause my pytest locally to apruptly terminate without any message, so I comment them out.

tomaarsen replied:
c612fc4516790b17604b5a1ff3d1b77c9ebbaf09

alvarobartt commented on src/argilla/client/feedback/training/frameworks/trl.py:
```diff
@@ -0,0 +1,383 @@
+        prepared_data=None,
+        model: str = None,
+        seed: int = None,
+    ):
```
```suggestion
    ) -> None:
```

alvarobartt commented on src/argilla/client/feedback/training/frameworks/trl.py:
```diff
@@ -0,0 +1,383 @@
+    _logger = logging.getLogger("ArgillaTRLTrainer")
+    _logger.setLevel(logging.INFO)
+
+    require_version("trl>=0.5.0")
```
Shouldn't we also include `transformers` and `torch`?

```suggestion
    require_version("transformers")
    require_version("torch")
    require_version("trl>=0.5.0")
```

alvarobartt replied:
Should be simpler and easier to define thanks to @plaguss PR at https://github.com/argilla-io/argilla/pull/3555

```python
require_dependencies(["trl>=0.5.0", "transformers", "torch"])
```

tomaarsen replied:
Sure. I didn't include these as `trl` requires transformers and torch already, so installing `trl>=0.5.0` should get you all set up.

tomaarsen replied:
We can use the `require_dependencies` later, I'll go with the triple `require_version` for now

alvarobartt commented on src/argilla/client/feedback/training/frameworks/trl.py:
```diff
@@ -0,0 +1,383 @@
+
+        self._transformers_model = None
+        self._transformers_tokenizer = None
+        self.device = "cpu"
+        if torch.backends.mps.is_available():
+            self.device = "mps"
+        elif torch.cuda.is_available():
+            self.device = "cuda"
```
Could we include a `use_cuda` and `use_mps` boolean flags as args? Maybe we don't want to use the device even if available

davidberenstein1957 replied:
I agree, this was to keep people from figuring out how to use `cuda` and `mps` but it can be overwritten via `update_config`. Perhaps it would be better to include it as one of the main args like `train_size`, however, this can be tackled in a different PR IMO. I believe @dvsrepo also suggested some other API changes.

tomaarsen replied:
I think it might be best to let users provide a `device` to the ArgillaTrainer init directly.

tomaarsen replied:
Let's track this in #3632 after this PR.

alvarobartt commented on src/argilla/client/feedback/training/frameworks/trl.py:
```diff
@@ -0,0 +1,383 @@
+        set_seed(self._seed)
+
+        if self._model is None:
+            self._model = "gpt2"
```
IMO we can assume the default is `gpt2-medium` instead of `gpt2`

```suggestion
            self._model = "gpt2-medium"
```

davidberenstein1957 replied:
For the other frameworks, I provided a default that works okay out of the box. Not sure if that would be reasonable here so perhaps it is better to go for a smaller model that allows for easy testing out of the box.

tomaarsen replied:
Sure, I'm cool with taking a stronger default here

alvarobartt commented on src/argilla/client/feedback/training/schemas.py:
```diff
@@ -39,12 +40,18 @@
 
 class TrainingData(ABC):
-    def _format_data(self, records):
+    def _format_data(self, dataset: "FeedbackDataset"):
```
```suggestion
    def _format_data(self, dataset: "FeedbackDataset") -> pd.DataFrame:
```

tomaarsen replied:
56f3ad5d84782f189a67aa1a0676e57144574def

alvarobartt commented on src/argilla/client/feedback/training/schemas.py:
```diff
@@ -128,30 +141,30 @@ def for_text_classification(
         Args:
             text (TextField): The TextField to use for training.
-            label (Union[RatingQuestion, LabelQuestion, MultiLabelQuestion, RatingQuestionUnification, LabelQuestionUnification, MultiLabelQuestionUnification]): _description_
+            label (Union[RatingQuestion, LabelQuestion, RankingQuestion, MultiLabelQuestion, RatingQuestionUnification, LabelQuestionUnification, MultiLabelQuestionUnification, RankingQuestionUnification]): _description_
```
As mentioned before, I think we can just remove the type-hints from the docstrings are those are automatically included when generating the docs if the type-hint is defined in the function arguments

alvarobartt replied:
Note that this comment also applies to the rest of the methods defined in this Python file

tomaarsen replied:
Sure, we can do that

tomaarsen replied:
1eff2089fbcf0ff6e21848da0781e600acf407fb

alvarobartt commented on tests/integration/client/feedback/training/test_trl.py:
```diff
@@ -0,0 +1,256 @@
+    "feedback_dataset_fields",
+    "feedback_dataset_questions",
+    "feedback_dataset_records",
+)
```
We should refactor the tests to use the pre-defined factories rather than those fixtures, but that can be tackled in a separate PR 👍🏻

tomaarsen replied:
These are normal fixtures, no idea why the weird `usefixtures` syntax is used, it's not needed at all.

tomaarsen replied:
7e2c28eb14688ab8898c370462a76e5783d826c7

davidberenstein1957 commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -376,11 +404,15 @@ def prepare_for_training(
                     lang = spacy.blank(lang)
```
```suggestion
                    lang = spacy.blank(lang) if len(lang)==2 else spacy.load(lang)
```

davidberenstein1957 replied:
FYI this suggestions does not really work ou-of-the box😅

tomaarsen replied:
I don't think I understand, this line is already wrapped in
```python
                if len(lang) == 2:
                    lang = spacy.blank(lang)
                else:
                    lang = spacy.load(lang)
```

davidberenstein1957 commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -47,37 +48,32 @@ def __init__(
             **load_kwargs: arguments for the rg.load() function.
         """
         self._dataset = dataset
         self._train_size = train_size
```
Maybe we should provide a warning if someone does provide a train_size?

tomaarsen replied:
Do you mean "does not"? I think it's mostly okay either way, sometimes just exclusively training with your data is fine-ish, especially for models where a loss doesn't really say much about how good a model is.

davidberenstein1957 commented on src/argilla/client/feedback/training/frameworks/trl.py:
```diff
@@ -0,0 +1,383 @@
+        if self._model is None:
+            self._model = "gpt2"
+
+        if isinstance(self._dataset, DatasetDict):
```
I think the train_size warning might need to go here and the concatenation of datasets can also be done here.

davidberenstein1957 commented on src/argilla/client/feedback/training/frameworks/trl.py:
```diff
@@ -0,0 +1,383 @@
+        self.training_args_kwargs.update(filter_allowed_args(PPOArgs.__init__, **kwargs))
+        self.trainer_kwargs.update(filter_allowed_args(self.trainer_cls.__init__, **kwargs))
+
+    def predict(self, text: Union[List[str], str], as_argilla_records: bool = True, **kwargs) -> None:
```
maybe we should allow for generative predictions?

tomaarsen replied:
Then I think a more logical next step is to add a separate generate method?

tomaarsen replied:
I'll make an issue for this

davidberenstein1957 commented on src/argilla/client/feedback/training/frameworks/trl.py:
```diff
@@ -0,0 +1,383 @@
+        """
+        self._transformers_model.save_pretrained(output_dir)
+        self._transformers_tokenizer.save_pretrained(output_dir)
+
```
should we add a `push_to_hub` method to the `transformer` based models? Similarly, we can do that [for spaCy](https://huggingface.co/docs/hub/spacy#sharing-your-models) too. IMO, we would ideally align this with a nicely generated card with a link to the dataset. @dvsrepo @alvarobartt , what do you think?

alvarobartt replied:
Sure, we can prepare a card once we are happy with the training interface on a separate PR if that's necessary

tomaarsen replied:
Generating a card is a very nice feature, but it would have to be custom made for each of the task-framework combinations, as they all store different information (e.g. logs) and in different ways.

tomaarsen replied:
I'll make an issue for the push_to_hub

tomaarsen replied:
#3634 #3633

davidberenstein1957 reviewed (approved):
Thanks @tomaarsen, looks great! Some minor API-design-related comments.  I'll make a separate PR for the docs.

tomaarsen committed:
Add Callable type hint

tomaarsen committed:
Merge branch 'feat/integration_trl' of https://github.com/argilla-io/argilla into integration/trl_sft

tomaarsen committed:
Merge branch 'develop' of https://github.com/argilla-io/argilla into integration/trl_sft

tomaarsen committed:
Add integration deps to pyproject.toml

tomaarsen committed:
Remove fetch_records from the ArgillaTrainer

But keep it in prepare_for_training

tomaarsen committed:
Reintroduce spacy tests

tomaarsen committed:
Also require transformers and torch

Co-authored-by: Alvaro Bartolome <alvarobartt@gmail.com>

tomaarsen committed:
Add Optional where needed

Co-authored-by: Alvaro Bartolome <alvarobartt@gmail.com>

tomaarsen committed:
Add missing None on init type hint

Co-authored-by: Alvaro Bartolome <alvarobartt@gmail.com>

tomaarsen committed:
Use gpt2-medium as the default instead of gpt2

Co-authored-by: Alvaro Bartolome <alvarobartt@gmail.com>

tomaarsen committed:
Add type hints for _format_data

tomaarsen committed:
Remove type hints from docstrings

In favour of only having them in the type hints

tomaarsen committed:
Remove unnecessary mark.usefixtures

tomaarsen committed:
Rename modelling to modeling

tomaarsen committed:
Add changelog entry

davidberenstein1957 committed:
Docs/trl docs revisited (#3636)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

I updated the docs for the `ArgillaTrainer` and the new tasks.

Closes #<issue_number>

**Type of change**

(Remember to title the PR according to the type of change)

- [X] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [X] `sphinx-autobuild` (read [Developer
Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation)
for more details)

**Checklist**

- [X] I added relevant documentation
- [X] I followed the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [X] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>
Co-authored-by: Alvaro Bartolome <alvaro@argilla.io>

davidberenstein1957 committed:
tests: added tests to test for wrong output formatting_func

davidberenstein1957 committed:
tests: fix test_prepare_for_training_text_classification_with_formatting_func by changing tuple type check
tests: fix test_prepare_for_training_sft by proper pydantic typing
